[feature][Ready for review] UI and models to register third-party OAuth2 clients

abought opened the pull request:
Moved from brianjgeiger/osf.io#86. Work in progress; opened for tracking purposes.

Partially implements requirements of #3323.

## Purpose
1. Provide models and UI to allowing users to register new OAuth2 applications
2. Demonstrate the first front-end views built on the new API. This entails adding support for cross-origin requests to the API and javascript calls.

## Summary of changes
1. Add a new database model (OAuth2App) to track information about user-registered OAuth2 applications.
2. Add new APIv2 endpoints that allow users to request and change information about their API applications (default intended restriction: only registered owner can make queries about an application)
    - `/v2/users/<user_id>/applications/` and `/v2/users/<user_id>/<client_id>/`
3. Add new UI at `/settings/applications` and `/settings/applications/<client_id>` allowing users to see list of currently registered applications, view detailed information, or create new.
4. Add (and configure) the django-cors-headers application, used by DRF to support OSF calls to the new API endpoints. (particularly for dev environment, where calls are explicitly cross-origin; may be of use later as more OSF-internal services need to coordinate with the API)

## To be done
- [x] UI / knockout code cleanup
- [x] Add styling to display of client id/ client secret on detail page
- [x] Verify that registered applications integrate with CAS server as expected
- [x] Add unit tests for JS/ viewmodel (and clean up stubs)
- [x] Write documentation for developers (how to connect using oauth) and find a place to host/ serve that documentation (see #4078)

Possibly post-MVP: 
- Add ability for user to revoke applications
- Revisit suppressing routes from swagger documentation
- Allow resetting of client secret via UI. May also invalidate tokens as part of this reset.
- Refactor Links field (using new `LinksFieldWithSelfLink` classes that will land in https://github.com/CenterForOpenScience/osf.io/pull/3683 )
- Restrict views that are accessible via the API (see #4057)

## References
- Issue brianjgeiger/osf.io#30
- Issue #3041 

## Testing notes
Local testing of this feature requires realCAS (for deleting an application/deauthorizing tokens) and apiV2 (for operations on application data).

Additionally, settings page views are currently only activated in dev mode, controlled via the `dev_only` decorator.

abought commented:
Committed initial version for review. The UI in particular may need some rework to meet OSF standards on the backend; the current code is torn between **seeming** to reload the page, and **actually** reloading the page, when a new application is registered. This is leading to some wonky architecture.

If the frontend looks like it would block others who need to work with registered applications, one option would be to split this PR into two: database models/ API, and frontend later. Let me know what would be best.

Squashing commits is also an option, due to the amount of rework involved early on.

brianjgeiger commented on api/base/settings/defaults.py:
```diff
@@ -129,6 +142,9 @@
 
 SWAGGER_SETTINGS = {
     'api_path': API_PATH,
+    #'exclude_namespaces': [
```
Remove these if you're not using them, but you can keep the TODO in.

brianjgeiger commented on api/users/permissions.py:
```diff
@@ -0,0 +1,15 @@
+from rest_framework import permissions
+
+
+class OwnerOnly(permissions.BasePermission):
```
@sloria, I wonder if it would make sense to consolidate all permissions to api.base.permissions. Prooobably not, but the reason I'm thinking so is if we do a refactoring of permissions, we would have them all in the same place and it'll be less likely to miss something. Also, permissions seem more like a central thing than a collection of things associated with the objects. Not a strong feeling, but a little niggling one, so I thought I'd see what you thought.

abought replied:
I'd support consolidating permissions in general; having a ton of scattered, highly specific implementations seems like a code smell. (especially if those rules are implemented differently across different API versions or views)

Along those lines, let me know if there are any changes I can make to the OAuth2 app model to support a more centralized and consistent set of permissions controls. Thanks.

brianjgeiger commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,271 @@
+    };
+
+// Load data from JSON
+ApplicationData.prototype.fromJSON = function (data) {
```
It might be nice to have this named symmetrically with the serialize function. Either this would be deserialize, or that would be toJSON

abought replied:
Renamed `serialize` to `toJSON` in the `ApplicationData` object.

brianjgeiger commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,271 @@
+        self.dataUrl = data.data.links.self;
+    });
+
+    request.fail(function (xhr, status, err) {
```
This looks fairly generic. Might want to go ahead and make it a function for osfHelpers. If you had a more specific growl error, then no, but as it stands, all that needs to happen is passing in a little context for Raven.

brianjgeiger commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,93 @@
+            </div>
+            <div id="app-fields">
+                <!-- TODO: Add revoke/ reset buttons -->
+                <form role="form">
```
It looks like you could bind this with: content and focus all of your data-binds. The only thing you're using outside of context that you aren't already calling absolutely is the dataurl. http://knockoutjs.com/documentation/with-binding.html

This is not necessarily good advice, knockout being what it is, but I figured I'd mention.

brianjgeiger commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,93 @@
+<%def name="javascript_bottom()">
+<script type="text/javascript">
+   ## Store mako variables on window so they are accessible from JS
+   ## modules. Not sure if this is a good idea.
```
It's not the best idea, but it's the way we do it until we come up with a better way or turn the site into a single-page app.

sloria replied:
Ha, I think my comments were copy-and-pasted...

But yeah, still not sure if it's a good idea. =)

abought replied:
Heehee. Indeed, the comments were copy-pasted verbatim. There's a lot of copy-paste redundancy in the profile page templates in general. Sorry about that.

brianjgeiger commented on website/templates/profile/oauth_app_list.mako:
```diff
@@ -0,0 +1,71 @@
+        <a href="${web_url_for('oauth_application_register')}" role="button" class="btn btn-default pull-right"><i class="fa fa-plus"></i> Register new application</a>
+        <div id="app-list">
+
+            <p data-bind="visible: (content().length == 0)">You have not registered any applications that can connect to the OSF.</p>
```
I'm a little concerned, in general, about the usability of this feature. So, on the one hand, it's a developer-specific feature, and it's being implemented in the way that it is everywhere else, so that's good. On the other, if you don't know what it is, will it get confused with Addons? We might want some help text to explain to people under what circumstances they would use this feature and try to make it super-clear that this isn't for the sorts of things that addons do. But without being really text heavy, 'cause nobody reads a bunch of text.

abought replied:
I totally agree that some concise, well-chosen help text would be useful. Most likely this would appear on the "Register new" page?

The sidebar navigation element link this page "Developer apps", but I wouldn't rely solely on the word "developer" to scare people away.

brianjgeiger commented:
Pass done. :penguin: Nothing terrible that I can see, so I think it won't be so bad. @sloria may likely see things that I don't, though, 'cause he's like that.

brianjgeiger commented:
Oh, and fix the failing JS Tests, naturally.

abought commented:
Updated PR for newest develop. Major changes since last review at 656b0a1a8940962149000f23beba6fe4c1ba6b80:  

- Added some basic frontend validation to "create application" form. If required fields are blank, user should be warned and prevented from submitting.
- Added 'cancel' button that goes from the application form to the list view
- Aligned OSF models with CAS models so that user-registered applications can be authenticated
- When an application is deactivated (deleted), it will now hit CAS to deactivate all outstanding tokens. (This feature may not work with FakeCAS). Unit tests have been updated to mock out CAS calls where relevant.

brianjgeiger commented on website/profile/views.py:
```diff
@@ -382,6 +383,39 @@ def user_notifications(auth, **kwargs):
+@must_be_logged_in
+def oauth_application_list(auth, **kwargs):
+    """Return app creation page with list of known apps"""
+    # TODO: Hardcoded URL
```
I think this is good now, yes?

abought replied:
It's still hardcoding the structure of the URL, even if it's not hardcoding the base route. I don't know if we have any system of consistently identifying routes that referenced by something other than views.

brianjgeiger replied:
You always have to reference by *something*, right? Whether it's by the view function or by the route or whatever, there still has to be a reference. In the case of API V2 routes, the only guarantee we give anyone, developers or users, that something is going to be consistent over time is that, if it's a public route, that the route is not going to change. So what's the problem with using the route this way?

Granted, this may not end up being a public route, but I'd rather not be moving around private routes a lot, either.

abought replied:
Fair enough, and pinning it to apiv2 base discourages the main change I was worried about. Made change locally and will push once all review and responses collected.

abought commented on website/project/model.py:
```diff
@@ -281,6 +283,72 @@ def node(self):
+    )
+
+    # Client ID and secret
+    client_id = fields.StringField(default=lambda: uuid.uuid4().hex,
```
Does modularodm provide a way to enforce uniqueness on a field? (all client IDs should be unique)

sloria replied:
`unique=True`, I believe.

abought commented on api/users/serializers.py:
```diff
@@ -50,3 +68,64 @@ class ContributorSerializer(UserSerializer):
+    client_id = ser.CharField(help_text="The client ID for this application (automatically generated)",
+                              read_only=True)
+
+    client_secret = ser.CharField(help_text="The client secret for this application (automatically generated)",
```
May want to lengthen the client secret slightly here per rfc6749. 40 characters would be a good lower bound. (2^160 possible combinations). This is 32 chars, and 2 of the fields in uuid are less random.

sloria commented on api/users/serializers.py:
```diff
@@ -1,7 +1,25 @@
+    elif isinstance(user, basestring):
+        try:
+            User.load(user)
+        except:
```
Don't use naked excepts. Also, `StoredObject.load` does not raise an exception if the record is not found--it returns `None`.

sloria commented on api/users/views.py:
```diff
@@ -86,3 +90,78 @@ def get_queryset(self):
+        except cas.CasHTTPError:
+            raise APIException("Could not revoke application auth tokens; please try again later")
+
+    def perform_update(self, serializer):
```
Is this an override?

abought replied:
Yes.

![screen shot 2015-07-09 at 11 04 50 am](https://cloud.githubusercontent.com/assets/2957073/8598726/5fb77924-262a-11e5-9e13-fa4f5cc7ad6e.png)

sloria commented on website/project/model.py:
```diff
@@ -281,6 +283,72 @@ def node(self):
+    name = fields.StringField(index=True, required=True)
+    description = fields.StringField(required=False)
+
+    create_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow,
```
For consistency with other models, let's go with `date_created`.

sloria commented on framework/auth/cas.py:
```diff
@@ -146,6 +146,20 @@ def _parse_profile(self, raw):
+        if resp.status_code == 204:
+            return True
+        else:
+            self.handle_error(resp)
```
Should this be `self._handle_error(resp)`?

sloria commented on website/project/decorators.py:
```diff
@@ -329,3 +330,15 @@ def wrapped(*args, **kwargs):
     return wrapper
+
+
+def dev_only(func):
```
Is this only for temporary use until API v2 is deployed?

abought replied:
Restricting certain views to dev-only was done at the request of @brianjgeiger ; pinging him for followup.

brianjgeiger replied:
Dev only is a handy decorator. I was planning on using it for the API playground page as well.

> On Jul 9, 2015, at 4:14 PM, Andy Boughton <notifications@github.com> wrote:
> 
> In website/project/decorators.py:
> 
> > @@ -329,3 +330,15 @@ def wrapped(*args, **kwargs):
> >  
> >      # Return decorator
> >      return wrapper
> > +
> > +
> > +def dev_only(func):
> Restricting certain views to dev-only was done at the request of @brianjgeiger ; pinging him for followup.
> 
> —
> Reply to this email directly or view it on GitHub.
>

sloria replied:
But why should the Oauth Applications UI views be dev-only? `dev_only` seems to be for views that are for developers of the OSF. Oauth Applications are for client applications.

abought replied:
Longterm I agree. I suspect this was a stopgap until we released the API, but will ping @brianjgeiger for followup on design rationale. 

Even if we keep this decorator in general, when should it stop being applied to settings pages for oauth platform application registration?

brianjgeiger replied:
Currently the API only works for situations where the API server is launched, which means it can be merged into staging and production without fear, but the settings page won't have that luxury. We are going to need it to be running on staging2 and develop well before it goes into production, which means we need a way of limiting it on Production.

Since staging2 is in develop mode, that means that we could have the api running with this UI but not have it show in Production. The downside is that staging is also in develop mode, which will complicate things. So we may need another flag. It might be that we could generalize this decorator so that you pass in the appropriate value/flag rather than having it know about a specific flag. That would make it potentially more usable for different circumstances. 

All that being said, when the API launches in production, the decorator should no longer be used for this particular item.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,317 @@
+
+'use strict';
+
+// TODO: Some of these may not be actually required. Dependency list ripped from profile view
```
Unused imports should be removed before this is merged.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,317 @@
+};
+
+// Serialize data for POST request
+ApplicationData.prototype.toJSON = function () {
```
For consistency, with other JS "models", use `serialize`/`deserialize`.

sloria commented on website/static/js/pages/profile-settings-applications-detail.js:
```diff
@@ -0,0 +1,7 @@
+'use strict';
+
+var viewModels = require('../apiApplication.js');
```
Omit the `.js` extension.

sloria commented on website/static/js/pages/profile-settings-applications-detail.js:
```diff
@@ -0,0 +1,7 @@
+var viewModels = require('../apiApplication.js');
+
+var ctx = window.contextVars;
+var apiApplication = new viewModels.ApplicationDetail('#app-detail', ctx.urls);
```
> Use .lowercase-and-dashes for class names and #camelCase for IDs.

https://cosdev.readthedocs.org/en/latest/style_guides/html_css.html

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,110 @@
+<%def name="title()">Application Detail</%def>
+<%def name="content()">
+<style type="text/css">
+    .border-box {
```
Is this overriding a Bootstrap class or is this a new class?

abought replied:
New class (see screenshot): box with top and bottom borders (no sides). Didn't see an existing class like this, but UI subject to revision.

<img width="698" alt="screen shot 2015-07-13 at 10 56 23 am" src="https://cloud.githubusercontent.com/assets/2957073/8652680/fb005f9c-294d-11e5-9933-eff3dadd1b6a.png">

sloria replied:
OK, then this class should go in `style.css`.

sloria replied:
Actually, since this is "non-critical" page-specific CSS, this can be put in its own CSS file, i.e. `project-settings-applications-detail.css`, which is loaded with `require()` in `project-settings.applications-detail.js`.

sloria replied:
This change has not yet been made.

sloria commented on website/templates/profile/addons.mako:
```diff
@@ -14,6 +14,9 @@
                 <li><a href="${ web_url_for('user_account') }">Account Settings</a></li>
                 <li><a href="#">Configure Add-ons</a></li>
                 <li><a href="${ web_url_for('user_notifications') }">Notifications</a></li>
+                % if dev_mode:
```
See https://github.com/CenterForOpenScience/osf.io/pull/3082/files#r34271090

sloria commented on website/templates/profile/settings.mako:
```diff
@@ -25,6 +25,9 @@
                 <li><a href="${ web_url_for('user_account') }">Account Settings</a></li>
                 <li><a href="${ web_url_for('user_addons') }">Configure Add-ons</a></li>
                 <li><a href="${ web_url_for('user_notifications') }">Notifications</a></li>
+                % if dev_mode:
```
See https://github.com/CenterForOpenScience/osf.io/pull/3082/files#r34271090

sloria commented on website/util/__init__.py:
```diff
@@ -69,7 +70,7 @@ def api_url_for(view_name, _absolute=False, _xml=False, *args, **kwargs):
                params=None,
                base_route=website_settings.API_DOMAIN,
-               base_prefix='',
+               base_prefix=api_settings.API_BASE,
```
It might make sense to do:

```python
def api_v2_url(..., base_prefix=None, ...):
    base_prefix = base_prefix or api_settings.API_BASE
```

This would allow us to modify api_settings.API_BASE in tests, if we ever needed to.

```python
    @classmethod
    def setUpClass(cls):
        api_settings.API_BASE = 'api/test`
```

For similar usage, see the search functions (https://github.com/CenterForOpenScience/osf.io/blob/b9495e80a8f68549b3f2683d0120b948645f4a1a/website/search/search.py#L58-61) and tests (https://github.com/CenterForOpenScience/osf.io/blob/b9495e80a8f68549b3f2683d0120b948645f4a1a/tests/test_elastic.py#L35)

Not quite sure if it's necessary, but it's something to consider.

abought replied:
That's an interesting alternate proposal. But it's worth noting that the keyword argument was originally designed specifically for use in tests (to keep them config-file-independent): 
https://github.com/CenterForOpenScience/osf.io/blob/develop/tests/test_utils.py#L66

What are the advantages of essentially monkey-patching the variable in this context?

sloria replied:
There may not be an advantage--I only thought of it because we've had to do it before. Like I said, just something to consider. If you decide it's not advantageous, keep it as is. Follow your :heart:

sloria commented on website/static/js/osfHelpers.js:
```diff
@@ -57,6 +57,35 @@ var apiV2Url = function (path, options){
+/*
+ * Perform an ajax request (cross-origin if necessary) that sends and receives JSON
+ */
+var ajaxJSON = function(method, url, options) {
```
Would be good to have some basic unit tests for this.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,317 @@
+
+ApplicationViewModel.prototype.fetch = applicationFetch;
+
+ApplicationViewModel.prototype.updateApplication = function () {
```
We are moving away from having too much HTTP work done in our ViewModels, as it's a poor separation of concerns.

A better pattern is to have the following entities:

- Model objects to store app data and any computeds
- A Client object that makes HTTP requests and returns Model
- A thin ViewModel which dispatches to the Client and updates the View accordingly

In theory, the Model and Client are reusable and framework-agnostic. In practice, the Model will contain `ko.observables` and `ko.computeds` so that we can easily integrate them with Knockout views. Even so, the pattern makes for better reusability and easier refactoring.

For a good example of this pattern, see https://github.com/CenterForOpenScience/osf.io/blob/b9495e80a8f68549b3f2683d0120b948645f4a1a/website/static/js/accountSettings.js

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,317 @@
+    var request = $osf.ajaxJSON('PATCH', url, {isCors: true, data: payload});
+
+    request.done(function (data) {
+        self.content().fromJSON(data.data);  // Update the data with what request returns- reflect server side cleaning
```
I'm not clear on what this is doing--is `content()` somehow being updated?

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,317 @@
+    };
+
+// Load data from JSON
+ApplicationData.prototype.fromJSON = function (data) {
```
This might make sense as a class method that returns an `ApplicationData` instance--I think this would clarify the confusion here: https://github.com/CenterForOpenScience/osf.io/pull/3082/files#r34273201

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,110 @@
+
+    <div class="col-sm-3">
+        <div class="panel panel-default">
+            <ul class="nav nav-stacked nav-pills">
```
This is duplicated from `profile/settings.mako` and is also out-of-date. Can you DRY this up?

sloria replied:
Hm, I see now that this is actually duplicated across all the profile settings sections.

If you're feeling ambitious, go ahead and refactor this out into a mako `include`. Otherwise, update this to be consistent with the new styling.

sloria commented on website/profile/views.py:
```diff
@@ -393,6 +394,37 @@ def user_notifications(auth, **kwargs):
+
+    try:
+        ApiOAuth2Application.find_one(Q('client_id', 'eq', client_id))
+    except:
```
No naked `excepts` please.

sloria commented on website/profile/views.py:
```diff
@@ -393,6 +394,37 @@ def user_notifications(auth, **kwargs):
+def oauth_application_register(auth, **kwargs):
+    """Register an API application: blank form view"""
+    submit_url = api_v2_url("users/{}/applications/".format(auth.user._id))  # POST request to this url
+    return {"submit_url": json.dumps(submit_url),
```
Would it make sense to pass the unencoded variables to the template, then do the "stringification" in the template?

abought replied:
Yeah, this needs to be brought up to spec with current best practices. (see conversion scheme here: https://github.com/CenterForOpenScience/osf.io/issues/2933#issuecomment-119757673 ) 

I'd have preferred to have less data transformation going on in templates, but... meh.

sloria commented:
Initial pass finished. :last_quarter_moon_with_face: 

Note for reviewers: I have only reviewed the code thus far--I have not yet checked for style/UI consistency.

abought commented on website/project/model.py:
```diff
@@ -261,6 +283,71 @@ def unreport_abuse(self, user, save=False):
+    """Registration and key for user-created OAuth API applications"""
+    _id = fields.StringField(
+        primary=True,
+        default=lambda: str(ObjectId())
```
Is this necessary: Some models use it, others don't.

abought commented:
UI refactor complete; now uses separate client/ data/ viewmodels + historyJS shim based replaceState functionality to avoid page reload.

This commit series also adds unit tests for ajaxJSON, and changes the way in which client secrets are generated. (better control over length)

brianjgeiger commented:
Save should not be an active button unless the information in the view is changed.

brianjgeiger commented:
Steps:
1) On an existing entry, delete one of the required URL fields and change focus from that field
2) See the required message appear
3) Add in a URL to the complaining field
4) Without losing focus on the field, click save

Expected:
The 'required' message would disappear, the entry would save and give a success message

Actual:
The 'required' message disappears, but there is no success message. Also, the entry does not save.

brianjgeiger commented:
When the change state is different from the saved state, trying to leave the page without saving should bring up a warning.

brianjgeiger commented:
Probably 'cancel' when there are no changes to be saved should be 'back' instead?

brianjgeiger commented:
I think the hover-over saying that delete is not reversible unnecessary given the 'are you sure' dialog box also says that. Let's drop the :information_source:

brianjgeiger commented:
In edit mode, it would make sense to have a delete button. That way, if someone did a very bad job naming things and is verifying that this one item is exactly the one that needs deleting, they don't have to find it in the list after to delete it.

brianjgeiger commented:
And make sure your button colors are all correct vs the style guide. Aside from that, ui functionality looks good to me. :penguin:

brianjgeiger commented:
Okay, maybe one more: I suspect the Register New Application button is not at the right level of wordiness according to our Style Guide.

abought commented:
Commits added.
- Changed behavior of save button: inactive when data is invalid or unchanged (rather than clicking to toggle toggle error display)
- Added delete button to detail view
- Warn user when changes would be discarded
- Change button colors and language for style guide
- Removed "info" tooltip on list view
- Language changes (delete--> deactivate and move language to `osfLanguage.js`)
- Improved permissions check and HTTP errors on app detail view (to match API restrictions)

Some slight visual oddities noted with knockout validators, incl URL validation #3149 .

Kept the cancel button named as is, mostly because the user *could* be coming from another page, and "back" was confusing.

abought commented:
To help test features like Client ID usage and token deactivation, a (wildly simplistic) client is here: https://github.com/abought/osf_oauth2_demo

abought commented:
Pinging @sloria to discuss next stage of code review.

Conducted in-person functional review with @brianjgeiger Mon Aug 17. Ran through registration, viewing, editing, and deactivation of tokens (using the sample client implementation).

Responded to UI review comments by @caneruguz on Phase 1 proposal/ gdoc as of Tues Aug 18.

lyndsysimon commented on api/applications/serializers.py:
```diff
@@ -0,0 +1,60 @@
+    owner = ser.CharField(help_text="The id of the user who owns this application",
+                          read_only=True,  # Don't let user register an application in someone else's name
+                          source='owner._id'
+                          )
```
Very minor style critique - this field has its closing brace on a line by itself, the others do not.

lyndsysimon commented on api/applications/serializers.py:
```diff
@@ -0,0 +1,60 @@
+    def absolute_url(self, obj):
+        return obj.absolute_url
+
+    def create(self, validated_data):
```
Should this not be a class method?

abought replied:
DRF dictates this: the method is part of the serializer (not the DB model). It overrides an inherited (instance) method from `rest_framework.serializers.BaseSerializer` that governs how API inputs are converted to database data.

lyndsysimon replied:
Gotchya. I've since been introduced a bit more thoroughly to DRF.

*shivers*

lyndsysimon commented on api/applications/views.py:
```diff
@@ -0,0 +1,89 @@
+
+from api.base.filters import ODMFilterMixin
+from api.base.utils import get_object_or_404
+from .permissions import OwnerOnly
```
Is there a reason we're mixing absolute and relative imports?

lyndsysimon commented on api/users/serializers.py:
```diff
@@ -2,7 +2,6 @@
 
 from api.base.serializers import JSONAPISerializer, LinksField, Link
 
-
```
I don't believe these newline shoudl have been removed - classes should be surrounded by two blank lines.

lyndsysimon commented:
Consider my first pass on this one complete. :star:

abought commented:
Thanks, Lyndsy. Added responses to functional review comments: "client secret" is now hidden by default on app detail page.

Also updated PR to address merge conflicts; passing to @sloria when he's ready.

sloria commented:
The page title differs from the other Settings pages:

![osf___notifications](https://cloud.githubusercontent.com/assets/2379650/9449764/c8a01a7e-4a72-11e5-88a1-e14cd74ae0ad.png)

![osf___developer_applications](https://cloud.githubusercontent.com/assets/2379650/9449802/ff39415a-4a72-11e5-9947-d36cc6b97b0b.png)

Also, "Developer Applications" should display as active when it is selected.

sloria commented:
The UI is inconsistent with the other settings pages.

Notifications settings:

![osf___notifications](https://cloud.githubusercontent.com/assets/2379650/9449849/49965efe-4a73-11e5-9b48-14d27a75dc3d.png)

Developer apps:

![osf___developer_applications](https://cloud.githubusercontent.com/assets/2379650/9449856/5c558862-4a73-11e5-9961-c548e1b9f586.png)

sloria commented on website/routes.py:
```diff
@@ -560,6 +561,27 @@ def make_url_map(app):
             OsfWebRenderer('profile/notifications.mako', trust=False),
         ),
 
+        Rule(
```
Ping @brianjgeiger and @JeffSpies for URL review.

brianjgeiger replied:
These look like what I would expect the web urls to be.

sloria commented:
These buttons are confusing:

![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9449912/acc813dc-4a73-11e5-86e7-814f35bad767.png)

Is there a need for both "Create" and "Save"? "Deactivate" probably doesn't belong with those buttons.

sloria commented:
Disregard previous comment re: buttons. A JS error caused the buttons to display incorrectly.

sloria commented:
After I click "Register New", I see red validation errors immediately, without saving the form:

![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9450009/3aee1b66-4a74-11e5-8d24-7c88c2b080b5.png)

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,418 @@
+        var request = $osf.ajaxJSON(method, url, {isCors: true, data: payload});
+
+        request.done(function (data) { // The server response will contain the newly created/updated record
+            ret.resolve(new ApplicationData(data.data));
```
Why not use `this.unserialize`?

sloria commented on website/static/js/pages/profile-settings-applications-list.js:
```diff
@@ -0,0 +1,7 @@
+
+var ctx = window.contextVars;
+var apiApplication = new viewModels.ApplicationsList('#appList', ctx.urls);
+ctx.apiApplication = apiApplication;
```
Why does this need to be exposed?

abought replied:
Originally used it for debugging purposes; I'll remove this line for production.

sloria commented on website/templates/include/profile/settings_navpanel.mako:
```diff
@@ -13,5 +13,10 @@
       <li class="${'active' if current_page == 'notifications' else ''}">
         <a href="${ '#' if current_page == 'notifications' else web_url_for('user_notifications') }">Notifications</a></li>
+
+      % if dev_mode:
```
I'm not sure this should be `dev_mode`. When `dev_mode = True`, it means that the application is in a non-production environment. 

I would think we'd have a separate settings to enable/disable developer apps.

brianjgeiger replied:
That's essentially what I'm looking for with this; only show them right now on a staging or local environment. The advantage of making a specific flag is that we'd be able to keep it from showing up on staging while still showing up on staging2, which is fine. The downside is it's another flag. One that will probably be going away soon.

I believe we had this discussion https://github.com/CenterForOpenScience/osf.io/pull/3082/files#diff-2a345ff09b5f57b5b1503ff0fa2060f3R353 and I'm happy to have the same solution for both.

abought replied:
Confirmed @brianjgeiger 's comment: this flag is here only until APIv2 is released into production.

sloria commented:
I will review this in greater depth after it has demoed and reviewed by @erinmayhood

abought commented:
Pushed responses to review. Main functional review with @erinmayhood concluded last week, and additional review for UI changes completed this morning.

sloria commented:
A few suggestions for improvement:

![osf___developer_applications](https://cloud.githubusercontent.com/assets/2379650/9523557/a0436a1c-4ca8-11e5-9137-00dbd4e869c7.png)

- The text is ambiguous and inconsistent with other "create" buttons on the site: use "New Application"
- The icon is unnecessary
- Lots of unnecessary whitespace next to the button

I suggest the following:

![osf___developer_applications](https://cloud.githubusercontent.com/assets/2379650/9523835/cb1f4ec6-4ca9-11e5-89f5-ecce6f0e733a.png)

ping @erinmayhood for feedback

abought commented:
The placement seems nice. The blue color is specified by the style guide:
https://centerforopenscience.github.io/osf-style/style.html#buttons

sloria commented:
The blue color is not appropriate when the button is placed in the panel header. See other usages:

![osf___realigned_stable_initiative](https://cloud.githubusercontent.com/assets/2379650/9524037/b8713176-4caa-11e5-98ea-ab8ecd7d9a13.png)


From GitHub:

![personal_access_tokens](https://cloud.githubusercontent.com/assets/2379650/9524049/cacd36f8-4caa-11e5-86da-be7bdffac7f7.png)

erinmayhood commented:
@sloria and @abought, yes, let's go with that suggested placement. Blue is in the style guide and the + icon is used semi-frequently for add actions currently in the OSF so I think it's ok. I like the New Application language - Andy if you make the switch the + icon will be necessary for the action to be clear.

abought commented:
I don't view the current state of the OSF as a thorough implementation of the style guide, so I'm trying to stick with stated guidelines over consistency. I'll defer to @caneruguz to clarify the rule about buttons in panels, and we can update the style guide if needed.

The style guide makes a fair case for [removing](https://centerforopenscience.github.io/osf-style/style.html#icons) the icon though. I'll change text and placement and remove the icon.

sloria commented:
This is also from the style guide:

![osf_style___ui_guidelines](https://cloud.githubusercontent.com/assets/2379650/9524265/0ae907b6-4cac-11e5-911a-c638400d7eed.png)

erinmayhood commented:
@abought Steve is right, I see this now. Let's remove the color for right now.

abought commented:
Pushed UI changes; for what it's worth, final form was discussed with @caneruguz before resubmitting.

I've also added an osf-style tracking issue to cover our discussion of the style guide. Below are screenshots WITH and WITHOUT any applications available.

<img width="990" alt="screen shot 2015-08-27 at 1 20 04 pm" src="https://cloud.githubusercontent.com/assets/2957073/9527736/c490e50a-4cbe-11e5-98dd-892e61a1594d.png">
<img width="983" alt="screen shot 2015-08-27 at 1 20 24 pm" src="https://cloud.githubusercontent.com/assets/2957073/9527735/c4905752-4cbe-11e5-8fe6-4a38c20b9a9c.png">

brianjgeiger commented:
Very nice.

sloria commented on api/applications/views.py:
```diff
@@ -0,0 +1,89 @@
+
+    serializer_class = ApiOAuth2ApplicationSerializer
+
+    renderer_classes = [renderers.JSONRenderer]  # Hide from web-browsable API tool
```
Why don't we want this to be browse-able?

abought replied:
This endpoint includes sensitive information (client secrets) that should not be exposed to users outside the OSF

In a separate followon PR, the plan is to remove this from swagger, restrict to the admin scope, and generally discourage external visibility.

sloria replied:
Thanks for the clarification.

sloria commented on framework/auth/cas.py:
```diff
@@ -149,6 +149,20 @@ def _parse_profile(self, raw):
             resp.attributes[attribute] = data['attributes'][attribute]
         return resp
 
+    def revoke_application_tokens(self, client_id, client_secret):
```
Can you add a test for this in `test_cas_authentication`?

abought replied:
I'd love to see more integration tests for the parts that rely on CAS, actually. I couldn't think of a meaningful test based on mocks, but discussion is heartily welcomed.

Issue: this function hits an endpoint on realCAS, which either succeeds or fails based on state stored in a service that is not used by travis test suite, or even some devs running locally.

sloria replied:
I was thinking more about unit tests, e.g. mocking out the response from the CAS server, then making sure that response is handled as expected.

sloria commented on tests/factories.py:
```diff
@@ -128,6 +126,17 @@ class TagFactory(ModularOdmFactory):
+
+    name = Sequence(lambda n: 'Example OAuth2 Application #{}'.format(n))
+
+    home_url = 'ftp://ftp.ncbi.nlm.nih.gov/'
```
Just to be safe, it'd be good to make these non-functioning URLs.

sloria commented:
There's a subtle flash of un-styled content (FOUC) on the create application page:

![fouc2](https://cloud.githubusercontent.com/assets/2379650/9550078/357cbf92-4d77-11e5-857b-3bf3bdd7cff0.gif)

Can you look into fixing this? The content should be hidden until the bindings have been applied.

sloria commented on website/oauth/models/__init__.py:
```diff
@@ -348,3 +355,68 @@ def handle_callback(self, response):
+                                   index=True)
+    client_secret = fields.StringField(default=generate_client_secret)
+
+    active = fields.BooleanField(default=True,  # Set to False if application is deactivated
```
For consistency with other models, let's call this `is_active`.

abought replied:
CAS checks this field name, and changing the field name in the OSF model would cause problems.

For now, I will make all the other changes, but am leaving this one deliberately untouched. It can be revisited separately when Michael is not backlogged.

sloria replied:
OK, that is fine.

abought replied:
A PR with the `active`--> `is_active field` name is teed up on top of Michael's feature/oauth-scopes branch: https://github.com/icereval/cas-overlay/pull/1

sloria commented on website/oauth/models/__init__.py:
```diff
@@ -348,3 +355,68 @@ def handle_callback(self, response):
+
+    # Properties used by Django and DRF "Links: self" field
+    @property
+    def absolute_api_v2_url(self):
```
Maybe we remove `v2` for future-proofing,  i.e. `absolute_api_url`?

Your call; I don't feel strongly about this.

brianjgeiger replied:
I think versioning it makes sense for future proofing. Who knows what v3 will be doing?

sloria replied:
When we move to v3, then the change would only need to be made to the implementation of this method, rather than having to make the change the name here *and* the API serialization code. This property effectively hides the implementation for getting this an Application's canonical API url.

abought replied:
In my understanding, API versioning is intended to allow more than one API to exist simultaneously. (allowing a gradual migration)

Using the same method for two APIs would make it difficult to support both API versions at the same time: there is no guarantee than APIv3 will have the same URL structure.

abought replied:
Will the links field refactor allow us to do away with the model method? If so I'll happily make the requisite change so that we can keep all serializer logic in one place. I support the broader goal there.

brianjgeiger replied:
Yes to the likelihood that we will have multiple APIs live simultaneously in the future.

I'm not sure about the answer to the second question as things stand. I don't believe that's going away right now. @pattisdr is anything you're doing ridding us for the need of this functionality?

pattisdr replied:
I haven't made any changes there.

sloria commented on website/profile/views.py:
```diff
@@ -374,6 +374,43 @@ def user_notifications(auth, **kwargs):
         'mailing_lists': auth.user.mailing_lists
     }
 
+@dev_only
```
Please add a `TODO` comment here to remove `dev_only` when APIv2 is in production.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,434 @@
+
+var ko = require('knockout');
+require('knockout.validation');
+require('knockout.punches');
```
Are you using knockout punches in your templates? Only import it if you think it's necessary or makes the template syntax significantly cleaner; otherwise, remove it.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,434 @@
+/*
+ * Fetch data about applications
+ */
+var ApplicationDataClient = oop.defclass({
```
Nice job with this.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,434 @@
+        this.apiListUrl = urls.apiListUrl;
+        // Set up data storage
+        this.appData = ko.observableArray();
+        this.sortByName = ko.computed(function () {
```
This can be a `pureComputed`.

sloria replied:
Since this acts as a property rather than a method, it might make sense to name this `sortedByName`.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,434 @@
+        }.bind(this));
+
+        // Prevent user from leaving page if there are unsaved changes
+        $(window).on('beforeunload', function () {
```
I don't believe this is necessary. These `beforeunload` checks make sense for forms that *update* existing data because it might not be clear to a user when changes are saved (e.g. some sites autosave on blur, some don't).

Since this for is for *creating* an application, the check is unnecessary; it is clear that the changes will only take effect when you click "Create".

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,434 @@
+            }.bind(this));
+        }
+    },
+    updateApplication: function () {
```
Methods that do async work should return a promise. In this case, you can just return `request`.

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,117 @@
+
+                <p><strong class="m-b-sm">Client secret</strong>
+                    <i class="fa fa-info-circle text-muted"
+                       data-bind="tooltip: {title:'The client secret is known only to you and the OSF. Do not display or expose this information.',
```
:+1:

sloria commented:
The position of the "Deactivate" button is intuitive.

![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9551604/32458fd0-4d80-11e5-87a3-1b1838fe7573.png)

The "Deactivate" button is unrelated to the form inputs; it's position should reflect that.

Let's put the Deactivation link in the top section (discussed in person with @erinmayhood):

![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9551491/857e31bc-4d7f-11e5-8a1d-a7dac9a6dda8.png)

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,117 @@
+    <div class="col-sm-9 col-md-7">
+        <div class="row">
+            <div class="col-sm-12">
+                <div class="breadcrumb"><i class="fa fa-angle-double-left"></i> <a href="${web_url_for('oauth_application_list')}">Return to list of registered applications</a></div>
```
Isn't the application list URL an attribute of the ViewModel? If so, then use Knockout to bind it to the `a` element. When possible, try not to mix KO and mako.

sloria commented on website/templates/profile/oauth_app_list.mako:
```diff
@@ -0,0 +1,69 @@
+            <div class="panel-heading clearfix">
+                <h3 class="panel-title" style="padding-bottom: 5px; padding-top: 5px;">Developer Applications</h3>
+                <div class="pull-right">
+                    <a href="${web_url_for('oauth_application_register')}" role="button" class="btn btn-sm btn-default">New application</a>
```
See previous comment re: mixing mako and KO.

sloria commented on website/templates/profile/oauth_app_list.mako:
```diff
@@ -0,0 +1,69 @@
+            </div>
+            <div class="panel-body">
+
+                <p>The OSF allows third-party web applications to connect to the OSF on behalf of other users, via the OAuth 2.0 web application flow.</p>
```
Remove the comma.

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,117 @@
+
+                    <div class="form-group">
+                        <label>Application description</label>
+                        <textarea class="form-control" placeholder="Application description is optional" data-bind="value: description"></textarea>
```
The placeholder can simply be "Optional"

sloria commented:
Pass finished.

sloria commented on framework/auth/cas.py:
```diff
@@ -149,6 +149,20 @@ def _parse_profile(self, raw):
+        url = furl.furl(self.BASE_URL)
+        url.path.segments.extend(('oauth2', 'revoke'))
+        url.args['client_id'] = client_id
+        url.args['client_secret'] = client_secret
```
Ping @icereval for review. Is putting the client secret in the query params the preferred approach?

abought replied:
Even though it's server-to-server, I'm going to say no to my own code on that one and change it. (though there is an ancillary question of whether CAS should respond to secrets in query parameters at all)

abought commented:
Responses to comments pushed. UI changes minimal and match screenshots proposed. Tests passing.

sloria commented on webpack.common.config.js:
```diff
@@ -34,6 +34,8 @@ var entry = {
     'files-page': staticPath('js/pages/files-page.js'),
     'profile-settings-page': staticPath('js/pages/profile-settings-page.js'),
     'profile-account-settings-page': staticPath('js/pages/profile-account-settings-page.js'),
+    'profile-settings-applications-list-page': staticPath('js/pages/profile-settings-applications-list.js'),
```
Minor, but the filenames should reflect the entry point names, i.e. add `-page` to the filenames.

sloria commented on website/templates/profile/oauth_app_list.mako:
```diff
@@ -0,0 +1,70 @@
+        webCreateUrl: ${ web_url_for('oauth_application_register') | sjson, n }
+    };
+</script>
+<script src=${"/static/public/js/profile-settings-applications-list-page.js" | webpack_asset}></script>
```
This only needs to be loaded if we're in `dev_mode`.

abought replied:
I agree. That's why the only view function linked to this template has the `dev_only` decorator. Am I missing something? 

If the user can get anywhere *near* seeing this page outside of dev mode, I want to know about it.

sloria replied:
Ah, that's right. OK, the conditional isn't needed here. Thanks for the clarification.

sloria commented on website/templates/profile/oauth_app_detail.mako:
```diff
@@ -0,0 +1,119 @@
+    });
+
+</script>
+<script src=${"/static/public/js/profile-settings-applications-detail-page.js" | webpack_asset}></script>
```
Same here; this only needs to be loaded in `dev_mode`.

abought replied:
Same as above: two view functions use this template (create + edit). Both already have the dev_only decorator.

sloria commented on website/static/js/osfLanguage.js:
```diff
@@ -70,5 +73,13 @@ module.exports = {
+        discardUnchanged: 'Are you sure you want to discard your unsaved changes?',
+        deactivateConfirm: 'Are you sure you want to deactivate this application for all users and revoke all access tokens? This cannot be reversed.',
+        deactivateError: 'Could not delete application. ' + REFRESH_OR_SUPPORT,
+        dataFetchError: 'Data not loaded. ' + REFRESH_OR_SUPPORT,
```
This error message is ambiguous:

![osf___developer_applications](https://cloud.githubusercontent.com/assets/2379650/9584078/058a01f6-4fdd-11e5-8275-63aa35bef630.png)

How about "Could not load Developer Applications at this time. Please refresh ..."

sloria commented:
I get a redundant error message when updating an application:

1. Create an application
2. Edit the application
3. In the form, make the Application Name (required) blank
4. Submit

![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9584484/5724053c-4fdf-11e5-8676-8b7d7047b4cb.png)

Data that doesn't pass client-side validation should not be sent to the server.

sloria commented:
![osf___application_detail](https://cloud.githubusercontent.com/assets/2379650/9584592/fea3b406-4fdf-11e5-9e41-247b14b47d9b.png)

When updating a field, the save button doesn't become active until a field loses focus.

We had decided in https://github.com/CenterForOpenScience/osf.io/issues/1035 not to disable submit buttons and show validation errors upon submitting. Can you make your forms consistent with that decision?

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,433 @@
+        if (!this.dirty()) {
+            this.visitList();
+        } else {
+            bootbox.confirm({
```
This modal is unnecessary. Remove it.

abought replied:
This modal was added in direct response to UI feedback by @brianjgeiger here: https://github.com/CenterForOpenScience/osf.io/pull/3082#issuecomment-125242396

Additionally, the style guide calls for modal confirmation when there is unsaved content or settings:
https://centerforopenscience.github.io/osf-style/ui.html#confirmations

I am unsure how to handle this suggestion.

brianjgeiger replied:
Is this not a thing we're doing any more?

sloria replied:
The style guide suggests:

> USE confirmation when:
> - The action will override an existing content AND the content can't be easily reverted.(e.g. Deleting files, wiki pages)

> DON'T USE confirmation when:
> - The action can easily be reverted (e.g instance settings changes that can be changed back)

This falls under the category of actions that can be reverted. Confirmation modals should be used sparingly, reserved mainly for destructive, irreversible actions. Overusing them not only interrupts user flow, but it also diminishes the protective value they're supposed to give (because the user becomes accustomed to ignoring them).

brianjgeiger replied:
:+1:

sloria replied:
I also looked at similar forms on GitHub, Dropbox, and Google; they generally do not use a modal for a "Cancel" action.

abought replied:
I appreciate the feedback, but this is a direct quote from the style guide:
```
USE confirmation when:
...
-    There is unsaved content or settings
```
If I change this, it should be accompanied by a direct and immediate change to the style guide. Incidentally, we use confirmation modals on other profile pages (discard changes).

sloria replied:
There seems to be (justifiable) ambiguity in the guidelines here. In this context, I think the modal should be removed.

However, @abought rightfully points that the other profile pages use the confirmation modals for "Discard"/"Cancel". If we were to remove them here, they should be removed on those pages as well. Though it's a small change, it may be outside the scope of this PR.

Pinging @erinmayhood on how to proceed.

sloria commented on website/static/js/apiApplication.js:
```diff
@@ -0,0 +1,433 @@
+            });
+        }
+    },
+    changeMessage: function (text, css, timeout) {
```
There is a ViewModel mixin in `changeMessage.js` with convenience methods for displaying messages. You can inherit from that to DRY this up.

abought commented:
> We had decided in #1035 not to disable submit buttons and show validation errors upon submitting. Can you make your forms consistent with that decision?

The button is not disabled because of validation errors. It is disabled because the contents of the form do not register as having changed since last save.

This change was made in response to https://github.com/CenterForOpenScience/osf.io/pull/3082#issuecomment-125239887 - opened by @brianjgeiger , author of #1035. I'm a bit unclear as to how to proceed here.

brianjgeiger commented:
Sorry, that was my fault. Go ahead and follow the suggestions from #1035 and leave it active. Make the UI respond as if something is submitted, but you don't necessarily need to make a request if the contents are not changed. It's clearly been far too long since I've done UI things.

sloria commented:
Pass finished.

abought commented:
Pushed responses to changes, but kept the modal in for now.

I think I got all the changes, but 160 comments into the thread it's all a bit of a blur.

Bear in mind that the list UI is planned for a significant refactor once we introduce token revocation UI. Comments always welcome, but it would help to indicate whether or not they are blocking issues.

Testing considerations for QA will focus on ability to register and delete activations. 

There may be a small window (1-5min) after registration where a new application can't immediately be disabled; this is due to a CAS implementation detail* and outside the scope of the UI here. 

* CAS periodically polls the database to get list of known applications. Error messages when clicking deactivate immediately after registration probably occur because application tokens can't be revoked until the application is known to exist.

erinmayhood commented:
@abought and @sloria confirmation modals can remain consistent for this release (so as Andy has it currently). Thanks.

sloria commented:
Client-side validation of the URL fields doesn't seem to be working.

![localhost_5000_settings_applications_d4ee60b5499d43dd917bd5bc43f241b9_](https://cloud.githubusercontent.com/assets/2379650/9636731/0585d84e-516b-11e5-8c63-b8123e2116ce.png)

abought commented:
> Client-side validation of the URL fields doesn't seem to be working.

See #3149 (JIRA [#OSF-3190]) for URL possibly relevant issues with URL validator used across multiple OSF pages.
